_posts/2019-01-13-dask-cudf-first-steps.md

```diff
@@ -0,0 +1,437 @@
+    in semantics.
+
+
+Alternatives
+------------
+
+We could also write a new dask-dataframe-style project around cuDF that was
+more able to deviate from the Pandas/Dask Dataframe API.  Until recently this
+has actually been the approach, and the
+[dask-cudf](https://github.com/rapidsai/dask-cudf) project did exactly this.
+This was probably a good choice early on to get started and prototype things.
+The project was able to implement a wide range of functionality including
+groupby-aggregations, joins, and so on using
+[dask delayed](https://docs.dask.org/en/latest/delayed.html).
+
+We're redoing this now on top of dask dataframe though, which means that we're
+losing some functionality that dask-cudf had before, but hopefully the
+functionality that we add now will be more stable and established on a firmer
+base.
+
+
+Status Today
+------------
+
+Today very little works, but what does is decently smooth.
+
+Here is a simple example that reads some data from many CSV files,
+picks out a column,
+and does some aggregations.
+
+```python
+from dask_cuda import LocalCUDACluster
+import dask_cudf
+from dask.distributed import Client
+
+cluster = LocalCUDACluster()  # runs on eight local GPUs
+client = Client(cluster)
+
+gdf = dask_cudf.read_csv('data/nyc/many/*.csv')  # wrap around many CSV files
+
+>>> gdf.passenger_count.sum().compute()
+184464740
+```
+
+*Also note, NYC Taxi ridership is significantly less than it was a few years ago*
+
+
+What I'm excited about in the example above
+-------------------------------------------
+
+-  All of the infrastructure surrounding the cuDF code, like the cluster setup,
+   diagnostics, JupyterLab environment, and so on, came for free, like any
+   other new Dask project.
+
+   Here is an image of my JupyterLab setup
+
+   <a href="https://matthewrocklin.com/blog/images/dask-cudf-environment.png">
+     <img src="https://matthewrocklin.com/blog/images/dask-cudf-environment.png"
+          alt="Dask + CUDA + cuDF JupyterLab environment"
+          width="70%">
+   </a>
+
+-  Our `df` object is actually just a normal Dask Dataframe.  We didn't have to
+   write new `__repr__`, `__add__`, or `.sum()` implementations, and probably
+   many functions we didn't think about work well today (though also many
+   don't).
+
+-  We're tightly integrated and more connected to other systems. For example if
+   we wanted to convert our dask-cudf-dataframe to a dask-pandas-dataframe then
+   we would use the standard cuDF function for this:
+
+   ```python
+   df = df.map_partitions(cudf.DataFrame.to_pandas)
+   ```
+
+   We don't have to write anything special like a separate `.to_dask_dataframe`
+   method nor handle other special cases.
+
+   The parallelism of Dask should be orthogonal to the choice of CPU/GPU.
+
+-  It's easy to switch hardware.  By avoiding separate `dask-cudf` code paths
+   it's easier to add cuDF to an existing Dask+Pandas codebase if we get a GPU,
```

Comment:
```suggestion
   it's easier to add cuDF to an existing Dask+Pandas codebase to run on GPUs,
```

Reply:
"if we get a GPU" sounds weird. Who is we, and why do we have to share one GPU? :)